e2e/test/E2ETestBase.cs

```diff
@@ -0,0 +1,47 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Threading;
+using Microsoft.ApplicationInsights.DataContracts;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace Microsoft.Azure.Devices.E2ETests
+{
+    /// <summary>
+    /// This class creates an instance of the logger for each test and logs the test result along with other useful information.
+    /// </summary>
+    public class E2ETestBase
+    {
+        protected TestLogging Logger { get; set; }
+        protected Stopwatch StopWatch { get; set; }
+        public TestContext TestContext { get; set; }
+
+        [TestInitialize]
+        public void TestInitialize()
+        {
+            StopWatch = new Stopwatch();
+            StopWatch.Start();
+            Logger = TestLogging.GetInstance(TestContext);
+            Logger.WriteLine($"Starting test - {TestContext.TestName}", SeverityLevel.Information);
+        }
+
+        [TestCleanup]
+        public void TestCleanup()
+        {
+            string testStatus = TestContext.CurrentTestOutcome.ToString();
+            StopWatch.Stop();
+            double elapsed = StopWatch.Elapsed.TotalSeconds;
+            var properties = new Dictionary<string, string>
+            {
+                { LoggingPropertyNames.TestRunTime, elapsed.ToString() }
+            };
+            Logger.WriteLine($"Finished test - {TestContext.TestName}", SeverityLevel.Information, properties);
+            // As this is not an application that keeps running, explicitly flushing is required to ensure we do not lose any logs.
+            // The recommendation from AI is to wait for 5 seconds after flushing.
+            Logger.Flush();
+            Thread.Sleep(5000);
```

Comment:
why do we need this sleep here? 
if we have 10 tests in a class, and we run all of them in parallel, `TestCleanup` would mark the end of processing for each of those tests, independently.

How would a "sleep" statement here have any impact?

Reply:
The logger is per test and the object will be gone once the test exits. Before exiting this thread, I want to ensure all the logs are flushed out. Application Insights does some processing in the background after you call Flush and they ask us to wait 5 seconds before exiting. Flush should be sufficient and it works but just to be safe I added this - https://devblogs.microsoft.com/premier-developer/application-insights-use-case-for-telemetryclient-flush-calls/ . This should not cause any issues, just 5 secs extra per test and they all run in parallel, so impact should be very low.

Reply:
To add more context - The way MsTest works is it creates and instance of the class per test, which implies an instance of logger per tests.

Reply:
So if I understand correctly, the sleep here is to hold on to the thread context for an additional 5secs, after Flush, so that the logs are incorporated completely?

Reply:
yes

Reply:
How much slower are tests?

Reply:
5 seconds total as they all run in parallel.

Reply:
I do have a dimension to monitor the time it takes as well so I could do some analysis on the pipeline. Plus we can compare run time in the pipeline with respect to what it takes today. I'm not sure if there is a better solution today. I did see that if we exit without flushing, we were losing logs.